### DISCUSSION TITLE
Help on setting up a new Linux project

### DISCUSSION THREAD

**Discussion by DeuSCaNoN** [uNetworking/uWebSockets#1597] [Q&A] [answered]
I'm using WSL and I successfully ran the _make_ command after downloading the source files for uWebSockets and uSockets so I know the environment is good!

However, I'm new to Linux makefiles (I've always been in Windows Visual Studio land) and I don't want to have my own code be living in the uWebSockets folder. I want to create my own project folder that includes/links the uWebSockets library, is there a guide on how to do that.

Example:
Path to uWebSocekts source code is
/mnt/j/Projects/LinuxLibraries/uWebSockets-20.37.0

Path to uSockets source code is 
/mnt/j/Projects/LinuxLibraries/uWebSockets-20.37.0/uSockets

and I have a new project folder
/mnt/j/Projects/NewProject/
with main.cpp that I wish to use uWebSockets in

**Comment by uNetworkingAB** [comment]
uWS is deliberately not tied to any one build system. In fact, you don't even need one - the whole project is just standards compliant header files with 1 single dependency (uSockets). So as long as you can grab the uSockets.a you are golden.

You can set up the project in any or all of the 50 billion different build systems you wish

**Reply by DeuSCaNoN** [reply]
So as someone who is coming into Linux build systems, do you have any recommendation for a first timer?

P.S.
I tried earlier today to try and get Visual Studios MSBuild Linux Console project to work but it kept failing in grabbing the uSocekts.a symbols.

**Reply by uNetworkingAB** [reply, marked as answer]
I personally don't use build systems (I'm thoroughly fed up with their complexity, bugs and abstractions and false sense of achieving anything at all). I use the most minimal 1-liner Makefile that just calls the compiler (knowing your compiler is all you need):

```
default:
 g++ -Ipath_to_uws_src path_to_uSockets.a myproject.cpp
```

Build systems are like, the most controversial topic and people are often times full on fanatically obsessed with their pick, that's why I refuse to use them.

**Reply by DeuSCaNoN** [reply]
Ok, that's a good enough of a start as any then, let me figure out the basics of makefiles and see if i can just make a really minimalistic build pipeline for myself. Thanks

(I'm mostly trying to port my existing project which has 6 header files and 5 .cpp files, so just need to understand the syntax of feeding those into the g++ compiler.)

**Reply by DeuSCaNoN** [reply]
Huh, I got this error when I tried running this

`g++ -I../LinuxLibraries/uWebSockets-20.37.0/src -I./Headers main.cpp ../LinuxLibraries/uWebSockets-20.37.0/uSockets/uSockets.a`
```
../LinuxLibraries/uWebSockets-20.37.0/src/App.h: In member function ‘uWS::TemplatedApp<SSL>&& uWS::TemplatedApp<SSL>::close()’:
../LinuxLibraries/uWebSockets-20.37.0/src/App.h:254:9: error: there are no arguments to ‘us_socket_context_close’ that depend on a template parameter, so a declaration of ‘us_socket_context_close’ must be available [-fpermissive]
  254 |         us_socket_context_close(SSL, (struct us_socket_context_t *) httpContext);
```
it said
```
../LinuxLibraries/uWebSockets-20.37.0/src/App.h:254:9: note: (if you use ‘-fpermissive’, G++ will accept your code, but allowing the use of an undeclared name is deprecated)`
```
so should I use -fpermissive ?
